### ISSUE TITLE
Allow dynamically enabling/disabling nonce as a source

### ISSUE THREAD

**Issue by devd** [w3c/webappsec-csp#88]

**Comment by mikewest** [comment]
Hello, and welcome to another bug you filed several months ago that I'm only just now looking into. :)

Basically same questions here: how would you expect this to work, and how would it work in browsers that don't understand whatever the answer to that question is?

**Comment by mikewest** [comment]
Also, can't you do this already by starting with `script-src 'nonce-whatever'` and then injecting `<meta http-equiv="CSP" content="script-src 'none'">`?

**Comment by devd** [comment]
no; right now whether or not eval will be allowed is a one-time deal. For backwards-compat, some things need eval, but right now I can't "stop the bleeding" easily. Any random JS library that gets introduced might add one more dependency on eval. 

If we could toggle (say) window.csp.allowEval then we could manually allow-disallow it. So, right before the known backwards compat uses, we allow it and then disallow it so that no one else is able to use eval. I am totally ok with the page breaking in a very horrible way if a new JS lib that I don't know about uses eval.

**Comment by mikewest** [comment]
1. Wouldn't JS libraries just start writing `eval(x)` as `csp.enableEval();eval(x);csp.disableEval();`? Or, more realistically, just putting `csp.enableEval();` at the top of every script? :)

2. You can get part of this by injecting a policy that disables script after "booting" your page. Do you need more granular support than that?

**Comment by devd** [comment]
not particularly worried about 1, but we can add a nonce to the call to make 1 very very difficult.

yeah, 2 kinda works but in a modern system, where you might do most of the work client-side (and additionally, sometimes you might do that work on the server side a la server side react), complexity balloons up :(

**Comment by devd** [comment]
err ..sorry, rereading your question in 2: yes, we do need more granular support. JS templates, some jquery calls, etc might need eval but vast majority don't. For examples, if we dynamically add JS templates to a page, 2 is unworkable.
